languages/python/polar/ffi.py

```diff
@@ -3,91 +3,119 @@
+        check_result(lib.polar_load(self.ptr, string, filename))
 
-    name: str
-    args: Sequence[Any]
-
-    def __str__(self):
-        return f'{self.name}({", ".join(self.args)})'
+    def new_query_from_str(self, query_str):
+        return Query(
+            check_result(lib.polar_new_query(self.ptr, to_c_str(query_str), 0))
+        )
 
-    def __eq__(self, other):
-        if not isinstance(other, Predicate):
-            return False
-        return (
-            self.name == other.name
-            and len(self.args) == len(other.args)
-            and all(x == y for x, y in zip(self.args, other.args))
+    def new_query_from_term(self, query_term):
+        return Query(
+            check_result(
+                lib.polar_new_query_from_term(self.ptr, ffi_serialize(query_term), 0)
+            )
         )
 
+    def next_inline_query(self):
+        q = lib.polar_next_inline_query(self.ptr, 0)
+        if is_null(q):
+            return None
+        else:
+            return Query(q)
 
-def check_result(result):
-    if result == 0 or is_null(result):
-        raise_error()
-    return result
+    def register_constant(self, name, value):
+        name = to_c_str(name)
+        value = ffi_serialize(value)
+        check_result(lib.polar_register_constant(self.ptr, name, value))
 
 
-def is_null(result):
-    return result == ffi.NULL
+class Query:
+    def __init__(self, ptr):
+        self.ptr = ptr
 
+    def __del__(self):
+        lib.query_free(self.ptr)
 
-def to_c_str(string):
-    return ffi.new("char[]", string.encode())
+    def call_result(self, call_id, value):
+        """Make an external call and propagate FFI errors."""
+        if value is None:
+            value = ffi.NULL
+        else:
+            value = ffi_serialize(value)
+        check_result(lib.polar_call_result(self.ptr, call_id, value))
 
+    def question_result(self, call_id, answer):
+        answer = 1 if answer else 0
+        check_result(lib.polar_question_result(self.ptr, call_id, answer))
 
-def ffi_serialize(value):
-    return to_c_str(json.dumps(value))
+    def application_error(self, message):
+        """Pass an error back to polar to get stack trace and other info."""
+        message = to_c_str(message)
+        check_result(lib.polar_application_error(self.ptr, message))
+
+    def next_event(self):
+        return QueryEvent(check_result(lib.polar_next_query_event(self.ptr)))
+
+    def debug_command(self, command):
+        check_result(lib.polar_debug_command(self.ptr, ffi_serialize(command)))
 
 
-def ffi_deserialize(string):
-    """Reconstruct Python object from JSON-encoded C string."""
-    try:
-        check_result(string)
-        return json.loads(ffi.string(string).decode())
-    finally:
-        if not is_null(string):
-            lib.string_free(string)
+class QueryEvent:
+    def __init__(self, ptr):
+        self.ptr = ptr
 
+    def get(self):
+        return ffi.string(self.ptr).decode()
 
```

Comment:
do we ever actually call del on the QueryEvent to make sure it gets freed?

Reply:
Oh good call. We don't but I thought the GC will do it?

Reply:
It wont always call __del__ when it goes out of scope like you’d want it to. Python does some weird object caching stuff. I think this came up before and we found some stackoverflow stuff about how __del__ shouldn’t be used for memory management like this. If you want an automatic thing you have to use a with statement but I think in this case this is fine we should just call del in the main event loop after handling an event.

Reply:
Added